mlflow/pyfunc/__init__.py

```diff
@@ -996,10 +996,10 @@ def _infer_spark_udf_return_type(model_output_schema):
 
 
 def _parse_spark_datatype(datatype: str):
-    from pyspark.sql.functions import udf
+    from pyspark.sql.types import _parse_datatype_string
 
-    return_type = "boolean" if datatype == "bool" else datatype
-    return udf(lambda x: x, returnType=return_type).returnType
+    datatype = "boolean" if datatype == "bool" else datatype
+    return _parse_datatype_string(datatype)
 
```

Comment:
Revert this to _parse_datatype_string because for spark connect, `udf(lambda x: x, returnType=return_type).returnType` returns `UnparsedDataType` instead of DataType, if we want to keep this function then we need to add additional logic for parsing it afterwards.